lang/c/src/schema.c

```diff
@@ -520,15 +523,61 @@ avro_schema_enum_symbol_append(const avro_schema_t enum_schema,
 	return 0;
 }
 
-int
-avro_schema_record_field_append(const avro_schema_t record_schema,
-				const char *field_name,
-				const avro_schema_t field_schema)
+static bool
+is_equal_schema_type(const avro_schema_t field_schema, const json_t *field_default)
+{
+        int  json_type = json_typeof(field_default);
+
+        switch (avro_typeof(field_schema)) {
+        case AVRO_STRING:
+        case AVRO_BYTES:
+                return json_type == JSON_STRING;
+
+        case AVRO_INT32:
+        case AVRO_INT64:
+                return json_type == JSON_INTEGER;
+
+        case AVRO_FLOAT:
+        case AVRO_DOUBLE:
+                return json_type == JSON_REAL;
+
+        case AVRO_BOOLEAN:
+                return (json_type == JSON_TRUE) || (json_type == JSON_FALSE);
+
+        case AVRO_NULL:
+                return json_type == JSON_NULL;
+
+        case AVRO_ARRAY:
+                return json_type == JSON_ARRAY;
+
+        case AVRO_RECORD:
+        case AVRO_ENUM:
+        case AVRO_FIXED:
+        case AVRO_MAP:
+        case AVRO_UNION:
+        case AVRO_LINK:
+                return json_type == JSON_OBJECT;
```

Comment:
Not necessarily, it depends on what the link points to.  Object isn't the right type for `enum` or `fixed` (those are strings), and `union` depends on what's in the union.  Look at the spec for all of the details, it tells you what kind of JSON is an appropriate default value for each kind of schema.